File: yoyo/bencode.py
# -*- coding:utf8 -*-
from functools import partial
from yoyo.error import (
    EncodeError,
    DecodeError,
)


def _encode_string(data):
    """benode编码字符串

    Args:
        data (str): 编码数据
    """
    return "{}:{}".format(len(data), data)


def _encode_int(data):
    """benode编码整数

    Args:
        data (int): 编码数据
    """
    return "i{}e".format(data)


def _encode_list(data):
    """benode编码列表

    Args:
        data (list): 编码列表
    """
    result = []
    for item in data:
        result.append(encode(item))
    return "l{}e".format("".join(result))


def _encode_dict(data):
    """benode编码列表

    Args:
        data (dict): 编码字典
    """
    result = []
    for key, val in data.items():
        result.append("{}{}".format(
            _encode_string(key),
            encode(val)))
    return "d{}e".format("".join(result))


def encode(data):
    """becode编码

    Args:
        data: 编码数据
    Returns:
        (str): 编码结果
    """
    if isinstance(data, str):
        return _encode_string(data)
    elif isinstance(data, int):
        return _encode_int(data)
    elif isinstance(data, list):
        return _encode_list(data)
    elif isinstance(data, dict):
        return _encode_dict(data)
    else:
        raise EncodeError("unknown type")


def _decode_string(data, start):
    """becode解码字符串

    Args:
        data (str): 要解码的字符串
        start (int): 解码位置
    Returns:
        (str): 解码的字符串
        (int): 下次解码位置
    """
    if start >= len(data) or not data[start].isdigit():
        return None, -1

    colon = data.find(':', start)
    if colon == -1:
        return None, -1

    s_length = data[start: colon]
    if not s_length.isdigit():
        return None, -1

    length = int(s_length)
    if length < 0:
        return None, -1

    index = colon + 1 + length
    if index > len(data):
        return None, -1

    return data[colon + 1: index], index


def _decode_int(data, start):
    """becode解码整数

    Args:
        data (str): 要解码的字符串
        start (int): 解码位置
    Returns:
        (int): 解码的整数
        (int): 下次解码位置
    """
    if start >= len(data) or data[start] != 'i':
        return None, -1

    e = data.find('e', start)
    if e == -1:
        return None, -1

    s_value = data[start + 1: e]
    try:
        value = int(s_value)
    except ValueError:
        return None, -1
    else:
        return value, e + 1


def _decode_list(data, start):
    """becode解码列表

    Args:
        data (str): 要解码的字符串
        start (int): 解码位置
    Returns:
        (list): 解码的列表
        (int): 下次解码位置
    """
    if start >= len(data) or data[start] != 'l':
        return None, -1

    index = start + 1
    result = []
    while index < len(data):
        if data[index] == 'e':
            break

        item, index = _decode_item(data, index)
        result.append(item)

    if index >= len(data):
        return None, -1

    return result, index + 1


def _decode_dict(data, start):
    """becode解码字典

    Args:
        data (str): 要解码的字符串
        start (int): 解码位置
    Returns:
        (dict): 解码的字典
        (int): 下次解码位置
    """
    if start >= len(data) or data[start] != 'd':
        return None, -1

    index = start + 1
    result = {}
    while index < len(data):
        if data[index] == 'e':
            break

        key, index = _decode_string(data, index)
        if key is None:
            return None, -1

        value, index = _decode_item(data, index)
        result[key] = value

    if index >= len(data):
        return None, -1

    return result, index + 1


def _decode_item(data, start):
    """becode解码字典

    Args:
        data (str): 要解码的字符串
        start (int): 解码位置
    Returns:
        解码的数据
        (int): 下次解码位置
    """
    for func in (_decode_int, _decode_string, _decode_list, _decode_dict):
        result, index = func(data, start)
        if result is not None:
            return result, index

    raise DecodeError("invalid bencode")

decode = partial(_decode_item, start=0)

File: dht/bencode.go
package dht

import (
	"bytes"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"unicode"
	"unicode/utf8"
)

// find returns the index of first target in data starting from `start`.
// It returns -1 if target not found.
func find(data []byte, start int, target rune) (index int) {
	index = bytes.IndexRune(data[start:], target)
	if index != -1 {
		return index + start
	}
	return index
}

func DecodeString(data []byte, start int) (result interface{}, index int, err error) {
	i := find(data, start, ':')
	if i == -1 {
		err = errors.New("':' not found when decode string")
		return
	}

	length, err := strconv.Atoi(string(data[start:i]))
	if err != nil {
		return
	}

	if length < 0 {
		err = errors.New("invalid length of string")
		return
	}

	index = i + 1 + length

	if index > len(data) || index < i+1 {
		err = errors.New("out of range")
		return
	}

	result = string(data[i+1 : index])
	return
}

func DecodeInt(data []byte, start int) (result interface{}, index int, err error) {
	index = find(data, start+1, 'e')
	if index == -1 {
		err = errors.New("':' not found when decode int")
		return
	}

	result, err = strconv.Atoi(string(data[start+1 : index]))
	if err != nil {
		return
	}
	index++

	return
}

func decodeItem(data []byte, start int) (result interface{}, index int, err error) {
	if start >= len(data) {
		err = errors.New("start out of range")
		return
	}

	var (
		decodeFunc     func([]byte, int) (interface{}, int, error)
		startCharacter = data[start]
	)
	switch startCharacter {
	case 'i':
		decodeFunc = DecodeInt
	case 'l':
		decodeFunc = DecodeList
	case 'd':
		decodeFunc = DecodeDict
	case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
		decodeFunc = DecodeString
	default:
		err = errors.New("invalid bencode when decode item")
		return
	}

	result, index, err = decodeFunc(data, start)
	return
}

func DecodeList(data []byte, start int) (result interface{}, index int, err error) {
	var item interface{}
	r := make([]interface{}, 0, 8)

	index = start + 1
	for index < len(data) {
		char, _ := utf8.DecodeRune(data[index:])
		if char == 'e' {
			break
		}

		item, index, err = decodeItem(data, index)
		if err != nil {
			return
		}
		r = append(r, item)
	}

	if index == len(data) {
		err = errors.New("'e' not found when decode list")
		return
	}
	index++

	result = r
	return
}

func DecodeDict(data []byte, start int) (result interface{}, index int, err error) {
	var item, key interface{}
	r := make(map[string]interface{})

	index = start + 1
	for index < len(data) {
		char, _ := utf8.DecodeRune(data[index:])
		if char == 'e' {
			break
		}

		if !unicode.IsDigit(char) {
			err = errors.New("invalid dict bencode")
			return
		}

		key, index, err = DecodeString(data, index)
		if err != nil {
			return
		}

		if index >= len(data) {
			err = errors.New("out of range")
			return
		}

		item, index, err = decodeItem(data, index)
		if err != nil {
			return
		}

		r[key.(string)] = item
	}

	if index == len(data) {
		err = errors.New("'e' not found when decode dict")
		return
	}
	index++

	result = r
	return
}

func Decode(data []byte) (result interface{}, err error) {
	if data == nil || len(data) == 0 {
		err = errors.New("decode data empty")
		return
	}

	result, _, err = decodeItem(data, 0)
	return
}

func EncodeString(data string) string {
	return fmt.Sprintf("%d:%s", len(data), data)
}

func EncodeInt(data int) string {
	return fmt.Sprintf("i%de", data)
}

func encodeItem(data interface{}) (result string, err error) {
	switch v := data.(type) {
	case string:
		result = EncodeString(v)
	case int:
		result = EncodeInt(v)
	case []interface{}:
		result, err = EncodeList(v)
	case map[string]interface{}:
		result, err = EncodeDict(v)
	default:
		err = errors.New("invalid type when encode result")
	}
	return
}

func EncodeList(data []interface{}) (result string, err error) {
	items := make([]string, len(data))

	for i, item := range data {
		items[i], err = encodeItem(item)
		if err != nil {
			return
		}
	}

	result = fmt.Sprintf("l%se", strings.Join(items, ""))
	return
}

func EncodeDict(data map[string]interface{}) (result string, err error) {
	items, i := make([]string, len(data)), 0

	for key, val := range data {
		item, e := encodeItem(val)
		if e != nil {
			err = e
			return
		}
		items[i] = fmt.Sprintf("%s%s", EncodeString(key), item)
		i++
	}

	result = fmt.Sprintf("d%se", strings.Join(items, ""))
	return
}

func Encode(data interface{}) (result string, err error) {
	result, err = encodeItem(data)
	return
}
